### ISSUE TITLE
Memory leaking with many TensorFlow models and warmup

### ISSUE THREAD

**Issue by whillas** [triton-inference-server/server#2237]
**Description**
We're loading ~150 efficientnet-b5 models on 3 RTX 2080 GPUs each with 11GB of memorry. After Triton reports that all models are loaded memory usage is as follows:
```
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 455.23.04    Driver Version: 455.23.04    CUDA Version: 11.1     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   1  GeForce RTX 208...  On   | 00000000:1A:00.0 Off |                  N/A |
| 18%   53C    P8    14W / 250W |   9133MiB / 11019MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   2  GeForce RTX 208...  On   | 00000000:67:00.0 Off |                  N/A |
| 18%   58C    P8    15W / 250W |   9131MiB / 11019MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   3  GeForce RTX 208...  On   | 00000000:68:00.0 Off |                  N/A |
| 18%   52C    P8    13W / 250W |   9446MiB / 11016MiB |      1%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
```
You can see the memory is around 83%

I then do a performance test on one of the models 
```
docker run --network host -it nvcr.io/nvidia/tritonserver:20.09-py3-clientsdk perf_client -u http://localhost:8000 -m my-efficientnet-b5 --shape inputs:1,456,456,3
```
and get `No valid requests recorded within time interval. Please use a larger time window.` the first time and lots of this repeating:
```
...
2020-11-10 04:31:08.089124: I tensorflow/core/common_runtime/bfc_allocator.cc:917] 1 Chunks of size 2975744 totalling 2.84MiB                                              
2020-11-10 04:31:08.089127: I tensorflow/core/common_runtime/bfc_allocator.cc:917] 1 Chunks of size 3440896 totalling 3.28MiB                                              
2020-11-10 04:31:08.089131: I tensorflow/core/common_runtime/bfc_allocator.cc:917] 1 Chunks of size 3580928 totalling 3.42MiB                                              
2020-11-10 04:31:08.089135: I tensorflow/core/common_runtime/bfc_allocator.cc:917] 49 Chunks of size 3735552 totalling 174.56MiB                                           
2020-11-10 04:31:08.089139: I tensorflow/core/common_runtime/bfc_allocator.cc:917] 48 Chunks of size 4194304 totalling 192.00MiB                                           
2020-11-10 04:31:08.089143: I tensorflow/core/common_runtime/bfc_allocator.cc:917] 1 Chunks of size 5201920 totalling 4.96MiB                                              
2020-11-10 04:31:08.089146: I tensorflow/core/common_runtime/bfc_allocator.cc:917] 194 Chunks of size 6291456 totalling 1.14GiB                                            
2020-11-10 04:31:08.089150: I tensorflow/core/common_runtime/bfc_allocator.cc:917] 1 Chunks of size 7313152 totalling 6.97MiB                                              
2020-11-10 04:31:08.089154: I tensorflow/core/common_runtime/bfc_allocator.cc:917] 1 Chunks of size 8388608 totalling 8.00MiB                                              
2020-11-10 04:31:08.089158: I tensorflow/core/common_runtime/bfc_allocator.cc:917] 1 Chunks of size 10485760 totalling 10.00MiB                                            
2020-11-10 04:31:08.089161: I tensorflow/core/common_runtime/bfc_allocator.cc:917] 1 Chunks of size 17238016 totalling 16.44MiB                                            
2020-11-10 04:31:08.089165: I tensorflow/core/common_runtime/bfc_allocator.cc:921] Sum Total of in-use chunks: 5.26GiB                                                     
2020-11-10 04:31:08.089169: I tensorflow/core/common_runtime/bfc_allocator.cc:923] total_region_allocated_bytes_: 8588886016 memory_limit_: 9610067968 available bytes: 1021181952 curr_region_allocation_bytes_: 8589934592                                                                                                                          
2020-11-10 04:31:08.089175: I tensorflow/core/common_runtime/bfc_allocator.cc:929] Stats:                                                                                  
Limit:                  9610067968                                                                                                                                         
InUse:                  5650828032                                                                                                                                         
MaxInUse:               7823081472                                                                                                                                         
NumAllocs:                   37929                                                                                                                                         
MaxAllocSize:           2176393216                                                                                                                                         
                                                                                                                                                                           
2020-11-10 04:31:08.090262: W tensorflow/core/common_runtime/bfc_allocator.cc:424] ****************__________________________________**************************************************    
```
I run it a second time and get results
```
*** Measurement Settings ***
  Batch size: 1
  Measurement window: 5000 msec
  Using synchronous calls for inference
  Stabilizing using average latency

Request concurrency: 1
  Client: 
    Request count: 155
    Throughput: 31 infer/sec
    Avg latency: 32393 usec (standard deviation 9025 usec)
    p50 latency: 28374 usec
    p90 latency: 45916 usec
    p95 latency: 47370 usec
    p99 latency: 49763 usec
    Avg HTTP time: 32497 usec (send/recv 1404 usec + response wait 31093 usec)
  Server: 
    Inference count: 185
    Execution count: 185
    Successful request count: 185
    Avg request latency: 28110 usec (overhead 55 usec + queue 49 usec + compute input 448 usec + compute infer 27517 usec + compute output 41 usec)
```
So model warm up problem I think and add this to all the model config files
```
model_warmup [{
    name: "start me UP!"
    batch_size: 1
    inputs[{
        key: "image_tensor"
        value: {
            data_type: TYPE_FP32
            dims: 1
            dims: 456
            dims: 456
            dims: 3
            random_data: true
        }
    }]
}]
```
Which results in the server never being ready and more of the same from above plus this error:
```
2020-11-10 03:18:37.623694: I tensorflow/core/common_runtime/bfc_allocator.cc:917] 1 Chunks of size 28656384 totalling 27.33MiB                                           
2020-11-10 03:18:37.623698: I tensorflow/core/common_runtime/bfc_allocator.cc:917] 1 Chunks of size 36474368 totalling 34.78MiB                                           
2020-11-10 03:18:37.623702: I tensorflow/core/common_runtime/bfc_allocator.cc:917] 1 Chunks of size 36781568 totalling 35.08MiB                                           
2020-11-10 03:18:37.623706: I tensorflow/core/common_runtime/bfc_allocator.cc:921] Sum Total of in-use chunks: 5.84GiB                                                    
2020-11-10 03:18:37.623710: I tensorflow/core/common_runtime/bfc_allocator.cc:923] total_region_allocated_bytes_: 9610067968 memory_limit_: 9610067968 available bytes: 0 
curr_region_allocation_bytes_: 17179869184                                                                                                                                
2020-11-10 03:18:37.623717: I tensorflow/core/common_runtime/bfc_allocator.cc:929] Stats:                                                                                 
Limit:                  9610067968                                                                                                                                        
InUse:                  6274873600                                                                                                                                        
MaxInUse:               8937364480                                                                                                                                        
NumAllocs:                   78889                                                                                                                                        
MaxAllocSize:           1083663616                                                                                                                                        
                                                                                                                                                                          
2020-11-10 03:18:37.624781: W tensorflow/core/common_runtime/bfc_allocator.cc:424] __________**************__________*___________******************************************************                                                                                                                                                             
2020-11-10 03:18:37.634098: I tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op.cc:733] Building a new TensorRT engine for TRTEngineOp_0 input shapes: [[1,456,456,3]$
2020-11-10 03:18:38.446444: E tensorflow/compiler/tf2tensorrt/utils/trt_logger.cc:41] DefaultLogger safeContext.cpp (124) - Cudnn Error in initializeCommonContext: 1 (Co$
ld not initialize cudnn, please check cudnn installation.)                                                                                                                
2020-11-10 03:18:38.449614: E tensorflow/compiler/tf2tensorrt/utils/trt_logger.cc:41] DefaultLogger safeContext.cpp (124) - Cudnn Error in initializeCommonContext: 1 (Could not initialize cudnn, please check cudnn installation.)                                                                                                                
2020-11-10 03:18:38.449758: W tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op.cc:748] Engine creation for TRTEngineOp_0 failed. The native segment will be used instead. Reason: Internal: Failed to build TensorRT engine                                                                                                                    
2020-11-10 03:18:38.466186: E tensorflow/stream_executor/dnn.cc:613] CUDNN_STATUS_INTERNAL_ERROR                                                                          
in tensorflow/stream_executor/cuda/cuda_dnn.cc(3323): 'cudnnConvolutionForward( cudnn.handle(), alpha, input_nd.handle(), input_data.opaque(), filter_nd.handle(), filter data.opaque(), conv.handle(), ToConvForwardAlgo(algorithm_desc), scratch_memory.opaque(), scratch_memory.size(), beta, output_nd.handle(), output_data.opaque())'         
2020-11-10 03:18:38.466498: I tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op.cc:733] Building a new TensorRT engine for TRTEngineOp_0 input shapes: [[1,456,456,3] 
E1110 03:18:38.466785 1 backend.cc:84] warmup error: Internal - 2 root error(s) found.                                                                                    
  (0) Internal: cuDNN launch failure : input shape([1,768,1,1]) filter shape([1,1,768,32])                                                                                
         [[{{node efficientnet-b5/model/blocks_14/se/conv2d/Conv2D}}]]                                                                                                    
  (1) Internal: cuDNN launch failure : input shape([1,768,1,1]) filter shape([1,1,768,32])                                                                                
         [[{{node efficientnet-b5/model/blocks_14/se/conv2d/Conv2D}}]]                                                                                                    
         [[logits/_1479]]                                                                                                                                                 
0 successful operations.                                                                                                                                                  
0 derived errors ignored.                                                                                                                                                 
2020-11-10 03:18:39.416942: I tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op.cc:733] Building a new TensorRT engine for TRTEngineOp_0 input shapes: [[1,456,456,3]$
2020-11-10 03:18:40.331318: E tensorflow/compiler/tf2tensorrt/utils/trt_logger.cc:41] DefaultLogger ../rtSafe/cublas/cublasWrapper.cpp (71) - cuBLAS Error in CublasWrapper: 3 (Could not initialize cublas. Please check CUDA installation.)                                                                                                       
2020-11-10 03:18:40.333542: E tensorflow/compiler/tf2tensorrt/utils/trt_logger.cc:41] DefaultLogger ../rtSafe/cublas/cublasWrapper.cpp (71) - cuBLAS Error in CublasWrapper: 3 (Could not initialize cublas. Please check CUDA installation.)                                                                                                       
2020-11-10 03:18:40.333653: W tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op.cc:748] Engine creation for TRTEngineOp_0 failed. The native segment will be used instead. Reason: Internal: Failed to build TensorRT engine                                                                                                                    

2020-11-10 03:18:40.359263: I tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op.cc:733] Building a new TensorRT engine for TRTEngineOp_0 input shapes: [[1,456,456,3]]
2020-11-10 03:18:41.315432: E tensorflow/compiler/tf2tensorrt/utils/trt_logger.cc:41] DefaultLogger safeContext.cpp (124) - Cudnn Error in initializeCommonContext: 1 (Could not initialize cudnn, please check cudnn installation.)                                                                                                                
2020-11-10 03:18:41.318223: E tensorflow/compiler/tf2tensorrt/utils/trt_logger.cc:41] DefaultLogger safeContext.cpp (124) - Cudnn Error in initializeCommonContext: 1 (Could not initialize cudnn, please check cudnn installation.)                                                                                                                
2020-11-10 03:18:41.318357: W tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op.cc:748] Engine creation for TRTEngineOp_0 failed. The native segment will be used instead. Reason: Internal: Failed to build TensorRT engine                                                                                                                    
2020-11-10 03:18:41.318994: I tensorflow/stream_executor/stream.cc:2100] [stream=0x7f839463c040,impl=0x7f83946cb780] did not wait for [stream=0x7f835535a300,impl=0x7f8394
6cb2c0]                                                                                                                                                                   
2020-11-10 03:18:41.319020: I tensorflow/stream_executor/stream.cc:5048] [stream=0x7f839463c040,impl=0x7f83946cb780] did not memcpy host-to-device; source: 0x7f6be19b2300
2020-11-10 03:18:41.319134: F tensorflow/core/common_runtime/gpu/gpu_util.cc:342] CPU->GPU Memcpy failed 
```
This last message is troubling

**Triton Information**
Official Triton container image 20.10

**To Reproduce**
Run 150 of this model across 3 GPUs (or as many as will take up ~83% of GPU memory) and add the model warm up to all the configs

Example of the model is here https://console.cloud.google.com/storage/browser/imagr-public-bucket/efficientnet-b5

**Expected behaviour**
For the models to be ready is a reasonable amount of time i.e. less than 15minutes

**Comment by mg515** [comment]
I don't think it's a memory leak, but from my experience the memory consumption goes up with larger batch size and input dimensions. After you do inference with a sample like that, e.g. [1, 456, 456, 3], the memory does not go back down, but Triton reserves it for further inference with the model. If you increase the batch size or the dimension size however, Triton will reserve more memory.

Warmup simply means Triton will do inference with one sample upon startup, therefore reserving memory for such inference, given batch and dimensions.

Long story short, your GPU memory is too small. Try to make a smaller input size just to see it work, or perhaps FP16 would be worth trying? Should take less memory than 32 precision.

Since you wrote in the title "with many TF models", does it work for you if you use only one model and one GPU?

**Comment by whillas** [comment]
Yes, this was done with a fixed size model [1, 456, 456, 3] inputs. So batches were of size 1. We also tried this with the optimisation 
```
optimization { execution_accelerators {
  gpu_execution_accelerator : [ {
    name : "tensorrt"
    parameters { key: "precision_mode" value: "FP16" }}]
}}
```
Which adds 10minutes to loading time and max'es out all 28 CPUs while doing it, i presume its compiling into an optimised version of TensorRT. We still get this problem.

I thought it was the batch dim. as well as before the input was [-1, 456, 456, 3] and was trying to use the max_batch_size to control it but it seems to be no-functional #2192 

This works fine for a single model on a single GPU, but then why do i need Triton for that scenario? As i stated in the initial bug report we have around 150 models and 3 GPUs (4 if we can get our custom Detectron2 running on Triton). We're at 83% memory utilisation which doesn't seem to b enough. I'm guessing that Triton (or the TF lib your using) tries to reserve memory for each individual model rather than sharing a pool of memory.

This worked fine in Triton v1.x My next step it to revert to container 20.07-v1 and see if (a) the models still work, and (b) if this memory issue arises.

**Comment by whillas** [comment]
We are running Triton with `--shm-size=1g --ulimit memlock=-1 --ulimit stack=67108864` as per the documentation but i can't find any information about what these values affect?

**Comment by deadeyegoodwin** [comment]
The latest documentation does not include those values. They a docker flags and are documented there.

**Comment by deadeyegoodwin** [comment]
TF allocates some GPU memory at load time for each model instance that you load. Then during each inference request it will allocate more GPU memory dynamically as needed for the inference computations (as mentioned the amount of memory allocated here will depend on the model and will typically increase with batch size). It appears that you are just putting too many models onto the GPU so that there is not enough memory remaining to use during the actual inference execution. Try reducing the number of model/model instances.
